src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -300,11 +279,94 @@ private ValueTask<Optional<TElement>> TryDequeueCoreAsync()
+        private static void AttachCancellation<T>(TaskCompletionSource<T> taskCompletionSource, CancellationToken cancellationToken)
+        {
+            _ = taskCompletionSource ?? throw new ArgumentNullException(nameof(taskCompletionSource));
+
+            if (!cancellationToken.CanBeCanceled || taskCompletionSource.Task.IsCompleted)
+                return;
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                taskCompletionSource.TrySetCanceled(cancellationToken);
+                return;
+            }
+
+            var tuple = new CancelableTaskCompletionSource<T>(taskCompletionSource, cancellationToken);
+            tuple.CancellationTokenRegistration = cancellationToken.Register(
+                static s =>
+                {
+                    var t = (CancelableTaskCompletionSource<T>)s!;
+                    t.TaskCompletionSource.TrySetCanceled(t.CancellationToken);
+                },
+                tuple,
+                useSynchronizationContext: false);
+
+            Debug.Assert(taskCompletionSource.Task.CreationOptions.HasFlag(TaskCreationOptions.RunContinuationsAsynchronously));
+            taskCompletionSource.Task.ContinueWith(
+                static (_, s) =>
+                {
+                    var t = (CancelableTaskCompletionSource<T>)s!;
+                    t.CancellationTokenRegistration.Dispose();
+                },
+                tuple,
+                CancellationToken.None,
+                TaskContinuationOptions.ExecuteSynchronously,
+                TaskScheduler.Default);
+        }
+
+        /// <summary>
+        /// A state object for tracking cancellation and a TaskCompletionSource.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned from a task.</typeparam>
+        /// <remarks>
+        /// We use this class so that we only allocate one object to support all continuations
+        /// required for cancellation handling, rather than a special closure and delegate for each one.
+        /// </remarks>
+        /// <seealso href="https://github.com/microsoft/vs-threading/blob/558f24c576cc620a00b20ed1fa90a5e2d13b0440/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs#L318-L372"/>
+        private class CancelableTaskCompletionSource<T>
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="CancelableTaskCompletionSource{T}"/> class.
+            /// </summary>
+            /// <param name="taskCompletionSource">The task completion source.</param>
+            /// <param name="cancellationToken">The cancellation token.</param>
+            internal CancelableTaskCompletionSource(TaskCompletionSource<T> taskCompletionSource, CancellationToken cancellationToken)
+            {
+                TaskCompletionSource = taskCompletionSource ?? throw new ArgumentNullException(nameof(taskCompletionSource));
```

Comment:
This is a `private` method, would `Debug.Assert` not `throw` here.

Reply:
➡️ This is nullable-enabled code; I just removed the checks